component/org.wso2.carbon.identity.casque.authenticator/src/main/java/org/wso2/carbon/identity/casque/authenticator/authenticator/radius/RadiusPacket.java

```diff
@@ -0,0 +1,396 @@
+                return att;
+            }
+        }
+        return null;
+    }
+
+    /**
+     * URL Encode the buffer.
+     * Replace + and /  with - and _.
+     *
+     * @param b the buffer.
+     */
+    private static void urlEncode(byte[] b) {
+
+        if (b != null) {
+            for (int i = 0; i < b.length; ++i) {
+                if (b[i] == '+') {
+                    b[i] = '-';
+                } else {
+                    if (b[i] == '/') {
+                        b[i] = '_';
+                    }
+                }
+            }
+        }
+    }
+
+    /**
+     * Parse the RADIUS Response Packet.
+     * Check the packet lengths and for a valid MD5 hash
+     * Get the response type, challenge and state values
+     * and package them in a RadiusResponse Object.
+     *
+     * @param packet the RADIUS Packet.
+     * @return the RadiusResponse Object.
+     */
+    static RadiusResponse parsePacket(DatagramPacket packet) {
+
+        int packetLength = packet.getLength();
+        if ((packetLength < 20) || (packetLength > 4096)) {
+            return new RadiusResponse(RadiusResponse.PACKET_LENGTH_ERROR);
+        }
+
+        byte[] tempBuffer = packet.getData();
+        int dataLength = (tempBuffer[2] & 0xff) * 256 + (tempBuffer[3] & 0xff);
+        if (packetLength < dataLength) {
+            return new RadiusResponse(RadiusResponse.PACKET_LENGTH_ERROR);
+        }
+
+        byte[] challenge = null;
+        byte[] state = null;
+        byte[] digest = new byte[16];
+        System.arraycopy(tempBuffer, 4, digest, 0, 16);
+        System.arraycopy(reqAuth, 0, tempBuffer, 4, 16);
+        MessageDigest md = getMD5();
+
+        if (md != null) {
+            md.update(tempBuffer, 0, dataLength);
+            md.update(CasqueConfig.radiusSecret);
+
+            byte[] digest2 = md.digest();
+
+            for (int i = 0; i < 16; i++) {
+                if (digest[i] != digest2[i]) {
+                    return new RadiusResponse(RadiusResponse.DIGEST_ERROR);
+                }
+            }
+
+            int offset = 20;
+            int index = 0;
+            byte[][] attributes = new byte[16][3];
+
+            while ((offset < dataLength) && (index < attributes.length)) {
+                attributes[index][0] = tempBuffer[offset];
+                attributes[index][1] = (byte) (offset + 2);
+                int l = (tempBuffer[offset + 1] & 0xff);
+                if (l < 2) {
+                    return new RadiusResponse(RadiusResponse.ATTRIBUTE_ERROR);
+                }
+                attributes[index][2] = (byte) (l - 2);
+                offset += l;
+                index++;
+            }
+
+            challenge = getAttribute(REPLY_MESSAGE, attributes, tempBuffer);
+            urlEncode(challenge);
+            state = getAttribute(STATE, attributes, tempBuffer);
+        }
+        return new RadiusResponse(tempBuffer[0], challenge, state);
+    }
+}
```

Comment:
Add a new line at the end.

Reply:
fixed with - https://github.com/wso2-extensions/identity-outbound-auth-casque-snr/pull/8